python/dgl/distributed/partition.py

```diff
@@ -1660,17 +1799,127 @@ def gb_convert_single_dgl_partition(
+        graph=part,
+        ntypes=ntypes,
+        etypes=etypes,
+        gpb=gpb,
+        part_meta=part_meta,
+        graph_formats=graph_formats,
+        store_eids=store_eids,
+        store_inner_node=store_inner_node,
+        store_inner_edge=store_inner_edge,
     )
-    torch.save(csc_graph, csc_graph_path)
+    rel_path = _save_graph_gb(part_config, part_id, csc_graph)
+    return rel_path
 
-    return os.path.relpath(csc_graph_path, os.path.dirname(part_config))
-    # Update graph path.
+
+def _convert_partition_to_graphbolt(
+    graph_formats,
+    part_config,
+    store_eids,
+    store_inner_node,
+    store_inner_edge,
+    n_jobs,
+    num_parts,
+):
+    # [Rui] DGL partitions are always saved as homogeneous graphs even though
+    # the original graph is heterogeneous. But heterogeneous information like
+    # node/edge types are saved as node/edge data alongside with partitions.
+    # What needs more attention is that due to the existence of HALO nodes in
+    # each partition, the local node IDs are not sorted according to the node
+    # types. So we fail to assign ``node_type_offset`` as required by GraphBolt.
+    # But this is not a problem since such information is not used in sampling.
+    # We can simply pass None to it.
+
+    # Iterate over partitions.
+    convert_with_format = partial(
+        convert_partition_to_graphbolt_multi_process,
+        part_config=part_config,
+        graph_formats=graph_formats,
+        store_eids=store_eids,
+        store_inner_node=store_inner_node,
+        store_inner_edge=store_inner_edge,
+    )
+    # Need to create entirely new interpreters, because we call C++ downstream
+    # See https://docs.python.org/3.12/library/multiprocessing.html#contexts-and-start-methods
+    # and https://pybind11.readthedocs.io/en/stable/advanced/misc.html#global-interpreter-lock-gil
+    rel_path_results = []
+    if n_jobs > 1 and num_parts > 1:
+        mp_ctx = mp.get_context("spawn")
+        with concurrent.futures.ProcessPoolExecutor(  # pylint: disable=unexpected-keyword-arg
+            max_workers=min(num_parts, n_jobs),
+            mp_context=mp_ctx,
+        ) as executor:
+            for part_id in range(num_parts):
+                rel_path_results.append(
+                    executor.submit(
+                        convert_with_format, part_id=part_id
+                    ).result()
+                )
+
+    else:
+        # If running single-threaded, avoid spawning new interpreter, which is slow
+        for part_id in range(num_parts):
+            rel_path = convert_with_format(part_id=part_id)
+            rel_path_results.append(rel_path)
+    part_meta = _load_part_config(part_config)
+    for part_id in range(num_parts):
+        # Update graph path.
+        part_meta[f"part-{part_id}"]["part_graph_graphbolt"] = rel_path_results[
+            part_id
+        ]
+
+    # Save dtype info into partition config.
```

Comment:
as this function is not called by direct saving from DGLGraph to GB, below metadata info is gone, not same as before. right?